File: 算法_js数据结构+python/python/数据结构与类型/stack.py
class Stack():
    def __init__(self):
        self.items = []
    def isEmpty(self):
        return self.items == []
    def push(self,element):
        self.items.append(element)
    def pop(self):
        return self.items.pop()
    def peek(self):
        return self.items[len(self.items)-1]
    def size(self):
        return len(self.items)
    def __str__(self):
        return str(self.items)


stack = Stack()



def rev(myStr):
    str1=''
    for i in myStr:
        stack.push(i)
    n=stack.size()
    print(stack)
    for i in range(n):
        str1 += stack.pop()
    print(str1)
rev('hello')

print(stack)


#进制转换
def sysConvert(x,base):
    stack = Stack()
    baseStr = ''
    digits = '0123456789ABCDEF'
    while x > 0:
        rem = x % base
        stack.push(rem)
        x = x // base
    while not stack.isEmpty():
        baseStr += str(digits[stack.pop()])
    return baseStr
print(sysConvert(10027,2))
print(sysConvert(10027,8))
print(sysConvert(10027,16))

File: 算法_js数据结构+python/js/数据结构与类型/Stack.js
//创建一个栈

function Stack(){
    let items = [];
    this.push = function(element){                       //添加元素到栈顶
        return items.push(element);
    };
    this.pop = function(){                              // 从栈顶移除元素
        return items.pop();
    };
    this.peek = function(){                            //返回到栈顶
        return items[items.length-1];
    };
    this.isEmpty = function(){                      //判断是否为空
        return items.length == 0;
    };
    this.clear = function(){                       //清空栈
        return items = [];
    };
    this.print = function(){                       //输出栈里的元素
        console.log(items.toString());
    };
    this.size = function(){                         //栈里元素个数
        return items.length;
    }
}
let stack=new Stack()
console.log(stack.isEmpty());
stack.push(5);
stack.push(2);
stack.push(8);
console.log(stack.peek());
stack.print();
console.log(stack.size())


//用栈进行二进制转换
function devideBy2(x){
    var remStack = new Stack(),
        rem,
        binaryString = '';
    while(x > 0){
        rem = Math.floor( x % 2 );
        remStack.push(rem);
        x = Math.floor( x / 2 );
    }
    while (!remStack.isEmpty()){
        binaryString += remStack.pop().toString();
    }
    return binaryString;
}

console.log(devideBy2(100))

//2,8,16进制
function devided(x,base){
    var remStack = new Stack(),
        rem,
        baseString = '',
        digits='0123456789ABCDEF';
    
    while(x > 0){
        rem = Math.floor( x % base);
        remStack.push(rem);
        x = Math.floor( x / base );
    }
    while (!remStack.isEmpty()){
        baseString += digits[remStack.pop()];
    }
    return baseString;
}
console.log(devided(10027,2))
console.log(devided(10027,8))
console.log(devided(10027,16))



